nav2_collision_monitor/src/source_base.cpp

```diff
@@ -0,0 +1,192 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/source_base.hpp"
+#include "nav2_collision_monitor/kinematics.hpp"
+
+#include <cmath>
+#include <exception>
+
+#include "tf2/convert.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_collision_monitor
+{
+
+SourceBase::SourceBase()
+: tf_buffer_(nullptr), source_type_(SOURCE_BASE), source_name_(""),
+  source_topic_(""), source_frame_id_(""), base_frame_id_(""),
+  transform_tolerance_(0.0), max_time_shift_(0.0)
+{
+}
+
+SourceBase::SourceBase(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string source_name,
+  const std::string base_frame_id,
+  const double transform_tolerance,
+  const double max_time_shift)
+: node_(node), tf_buffer_(tf_buffer), source_type_(SOURCE_BASE), source_name_(source_name),
+  source_topic_(""), source_frame_id_(""), base_frame_id_(base_frame_id),
+  transform_tolerance_(transform_tolerance), max_time_shift_(max_time_shift)
+{
+  auto node_sptr = node_.lock();
+  if (node_sptr) {
+    RCLCPP_INFO(node_sptr->get_logger(), "Creating SourceBase");
+  }
+}
+
+SourceBase::~SourceBase()
+{
+  auto node = node_.lock();
+  if (node) {
+    RCLCPP_INFO(node->get_logger(), "Destroying SourceBase");
+  }
+
+  tf_buffer_.reset();
+}
+
+bool SourceBase::getParameters()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  try {
+    nav2_util::declare_parameter_if_not_declared(
+      node, source_name_ + ".topic",
+      rclcpp::ParameterValue("scan"));  // Set deafult topic for laser scanner
+    source_topic_ = node->get_parameter(source_name_ + ".topic").as_string();
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic source parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+SourceType SourceBase::getSourceType()
+{
+  return source_type_;
+}
+
+std::string SourceBase::getSourceTypeStr()
+{
+  if (source_type_ == SOURCE_BASE) {
+    return "base";
+  } else if (source_type_ == SCAN) {
+    return "scan";
+  } else {  // source_type_ == POINTCLOUD
+    return "pointcloud";
+  }
+}
+
+bool SourceBase::getTransform(
+  const std::string to_frame,
+  const std::string from_frame,
+  tf2::Transform & tf2_transform)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!tf_buffer_) {
```

Comment:
Remove check, this should always be true

Reply:
Done in https://github.com/ros-planning/navigation2/pull/2982/files#diff-63f43bf4a53d7cc45e70a76a2ddc967ad118a455106569b3144e26ee3e867437R123